File: easy/binary_number_to_integer_linked_list/binary_number_to_integer_linked_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        return

    def push(self, val):
        new_node = Node(val)
        curr = self
        while curr.next != None:
            curr = curr.next
        curr.next = new_node
        return

    def print(self):
        curr = self
        while curr != None:
            print(curr.val, end=" -> ")
            curr = curr.next
        print()
        return


class Solution(object):
    def getDecimalValue(self, head):
        """
        :type head: ListNode
        :rtype: int
        """
        binary_number_str = ""
        while head != None:
            binary_number_str += str(head.val)
            head = head.next

        return int(binary_number_str, 2)


s = Solution()

print("Solution 1")
head = Node(1)
head.push(0)
head.push(1)
head.print()
print("Number : ", s.getDecimalValue(head))

print("Solution 2")
head = Node(0)
head.print()
print("Number : ", s.getDecimalValue(head))

print("Solution 3")
head = Node(1)
head.push(0)
head.push(0)
head.push(1)
head.push(0)
head.push(0)
head.push(1)
head.push(1)
head.push(1)
head.push(0)
head.push(0)
head.push(0)
head.push(0)
head.push(0)
head.push(0)
head.print()
print("Number : ", s.getDecimalValue(head))

print("Solution 4")
head = Node(1)
head.print()
print("Number : ", s.getDecimalValue(head))

print("Solution 5")
head = Node(0)
head.push(0)
head.print()
print("Number : ", s.getDecimalValue(head))

File: easy/binary_number_to_integer_linked_list/binary_number_to_integer_linked_list.go
package main

import (
	"fmt"
	"strconv"
)

type Node struct {
	Val  int
	Next *Node
}

func (n *Node) push(val int) {
	newNode := &Node{Val: val}
	curr := n
	for curr.Next != nil {
		curr = curr.Next
	}
	curr.Next = newNode
	return
}

func (n *Node) print() {
	curr := n
	for curr != nil {
		fmt.Printf(" %d ->", curr.Val)
		curr = curr.Next
	}
	fmt.Println()
	return
}

func getDecimalValue(head *Node) int {
	binaryNumberStr := ""
	curr := head
	for curr != nil {
		binaryNumberStr += strconv.Itoa(curr.Val)
		curr = curr.Next
	}

	numberInt, err := strconv.ParseInt(binaryNumberStr, 2, 64)
	if err != nil {
		return 0
	}
	return int(numberInt)
}

func main() {

	fmt.Println("Solution 1")
	head := &Node{Val: 1}
	head.push(0)
	head.push(1)
	head.print()
	fmt.Println("Number : ", getDecimalValue(head))

	fmt.Println("Solution 2")
	head = &Node{Val: 0}
	head.print()
	fmt.Println("Number : ", getDecimalValue(head))

	fmt.Println("Solution 3")
	head = &Node{Val: 1}
	head.push(0)
	head.push(0)
	head.push(1)
	head.push(0)
	head.push(0)
	head.push(1)
	head.push(1)
	head.push(1)
	head.push(0)
	head.push(0)
	head.push(0)
	head.push(0)
	head.push(0)
	head.push(0)
	head.print()
	fmt.Println("Number : ", getDecimalValue(head))

	fmt.Println("Solution 4")
	head = &Node{Val: 1}
	head.print()
	fmt.Println("Number : ", getDecimalValue(head))

	fmt.Println("Solution 5")
	head = &Node{Val: 0}
	head.push(0)
	head.print()
	fmt.Println("Number : ", getDecimalValue(head))
}
